File: src/agent-sec-core/hermes-plugin/src/capabilities/code_scan.py
"""Code-scan capability — scans terminal/execute_code via agent-sec-cli."""

from __future__ import annotations

import json
import logging

from ..cli_runner import call_agent_sec_cli
from .base import AgentSecCoreCapability

logger = logging.getLogger("agent-sec-core")

# Mapping: tool_name -> (args_key, language)
_TOOL_LANGUAGE_MAP = {
    "terminal": ("command", "bash"),
    "execute_code": ("code", "python"),
}


class CodeScanCapability(AgentSecCoreCapability):
    """Security capability that scans code before execution.

    Intercepts pre_tool_call for 'terminal' and 'execute_code' tools,
    sends the code to agent-sec-cli scan-code, and blocks on deny/warn verdicts.
    """

    id = "code-scan"
    name = "Code Scanner"

    def _on_register(self, config: dict) -> None:
        """Read code-scan specific config."""
        self._enable_block = config.get("enable_block", False)

    def get_hooks_define(self) -> dict:
        return {"pre_tool_call": self._on_pre_tool_call}

    def _on_pre_tool_call(self, tool_name, args, **kwargs):
        """Hook handler: scan terminal/execute_code for security risks."""
        # 1. Only intercept known tools
        tool_info = _TOOL_LANGUAGE_MAP.get(tool_name)
        if tool_info is None:
            return None

        args_key, language = tool_info

        # 2. Extract code content
        code = (args or {}).get(args_key, "").strip()
        if not code:
            return None

        # 3. Call agent-sec-cli scan-code
        result = call_agent_sec_cli(
            ["scan-code", "--code", code, "--language", language],
            timeout=self._timeout,
        )

        # 4. Parse result (fail-open on errors)
        if result.exit_code != 0:
            logger.warning(
                f"[agent-sec-core] {self.id} agent-sec-cli exit_code={result.exit_code}, fail-open tool={tool_name} code={code[:120]}"
            )
            return None

        try:
            scan = json.loads(result.stdout)
        except (json.JSONDecodeError, ValueError):
            logger.warning(
                f"[agent-sec-core] {self.id} agent-sec-cli returned invalid JSON, fail-open tool={tool_name} code={code[:120]}"
            )
            return None

        verdict = scan.get("verdict", "pass")

        # warn and deny are separate branches (coding convention), same behavior
        if verdict == "deny":
            msg = self._format_message(scan)
            logger.warning(
                f"[agent-sec-core] {self.id} DENY tool={tool_name} code={code[:120]} | {msg}"
            )
            if self._enable_block:
                return {"action": "block", "message": msg}
            return None

        if verdict == "warn":
            msg = self._format_message(scan)
            logger.warning(
                f"[agent-sec-core] {self.id} WARN tool={tool_name} code={code[:120]} | {msg}"
            )
            if self._enable_block:
                return {"action": "block", "message": msg}
            return None

        if verdict == "pass":
            logger.info(
                f"[agent-sec-core] {self.id} PASS tool={tool_name} code={code[:120]}"
            )
            return None

        # verdict == "error" — scanner itself failed, fail-open
        if verdict == "error":
            logger.warning(
                f"[agent-sec-core] {self.id} agent-sec-cli returned verdict=error, fail-open tool={tool_name} code={code[:120]}"
            )
            return None

        # unknown verdict — defensive fallback, fail-open
        logger.warning(
            f"[agent-sec-core] {self.id} UNKNOWN verdict={verdict} tool={tool_name} code={code[:120]}"
        )
        return None

    def _format_message(self, scan: dict) -> str:
        """Format scan-code result into a human-readable block message."""
        summary = scan.get("summary", "")
        findings = scan.get("findings", [])
        lines = [
            (
                f"[agent-sec-core] {summary}"
                if summary
                else "[agent-sec-core] Code scan blocked"
            )
        ]
        for f in findings:
            rule_id = f.get("rule_id", "?")
            desc = f.get("desc_zh") or f.get("desc_en", "")
            lines.append(f"  - {rule_id}: {desc}")
        return "\n".join(lines)

File: src/agent-sec-core/openclaw-plugin/src/capabilities/code-scan.ts
import type { SecurityCapability } from "../types.js";
import { buildTraceContext, callAgentSecCli } from "../utils.js";

export const codeScan: SecurityCapability = {
  id: "scan-code",
  name: "Code Scanner",
  hooks: ["before_tool_call"],
  register(api) {
    const cfg = (api.pluginConfig as Record<string, any>) ?? {};
    const requireApprovalEnabled = cfg.codeScanRequireApproval === true;

    api.on("before_tool_call", async (event: any, ctx: any) => {
      try {

        // 只拦截 shell 类工具
        const command = extractCommand(event);
        if (!command) {
          return undefined;
        }

        const result = await callAgentSecCli(
          ["scan-code", "--code", command, "--language", "bash"],
          { timeout: 10000, traceContext: buildTraceContext(event, ctx) },
        );

        if (result.exitCode !== 0) {
          return undefined;
        }

        const scanResult = JSON.parse(result.stdout);
        const verdict = scanResult.verdict;
        const findings = scanResult.findings ?? [];

        // Self-protect: force block if the command would disable this plugin
        const selfProtectFinding = findings.find(
          (f: any) => f.rule_id === "shell-self-protect-openclaw",
        );
        if (selfProtectFinding) {
          const msg = `[agent-sec-core] 自我保护：该命令将禁用 agent-sec 安全插件。如果您确实需要禁用，请手动执行以下命令：\n\n  ${command}\n\n出于安全原因，AI agent 无法执行此操作。`;
          api.logger.warn(`[scan-code] SELF-PROTECT block — ${command}`);
          return { block: true, blockReason: msg };
        }

        if (verdict === "pass" || findings.length === 0) {
          api.logger.info(`[scan-code] ✅ pass — allowing command`);
          return undefined;
        }

        // 构建提示信息（与 cosh hook 的 msg 格式一致）
        const descs = findings.map((f: any) => `- ${f.desc_zh}`);
        const msg = `[code-scanner] Detected ${findings.length} issue(s):\n${descs.join("\n")}\n\nCommand: ${command}`;

        if (verdict === "deny") {
          api.logger.warn(`[scan-code] DENY (requireApproval=${requireApprovalEnabled}) — ${msg}`);
          if (requireApprovalEnabled) {
            return {
              requireApproval: {
                title: "Code Scanner Security Warning",
                description: msg,
                severity: "warning" as const,
              },
            };
          }
          return undefined;
        }

        if (verdict === "warn") {
          api.logger.warn(`[scan-code] WARN (requireApproval=${requireApprovalEnabled}) — ${msg}`);
          if (requireApprovalEnabled) {
            return {
              requireApproval: {
                title: "Code Scanner Security Warning",
                description: msg,
                severity: "warning" as const,
              },
            };
          }
          return undefined;
        }

        return undefined;
      } catch (err) {
        return undefined; // crash ≠ threat → allow
      }
    });
  },
};

/** 从 event 中提取 shell 命令，无法提取则返回 undefined */
function extractCommand(event: { toolName: string; params: Record<string, unknown> }): string | undefined {
  // OpenClaw 唯一的 shell 执行工具是 exec，参数字段为 command
  // 参考: https://docs.openclaw.ai/tools/exec
  if (event.toolName !== "exec") return undefined;
  const cmd = event.params.command;
  if (typeof cmd !== "string" || !cmd.trim()) return undefined;
  return cmd;
}
